File: Hush/HotKeys.m
#import "HotKeys.h"

@implementation HotKeys

static OSStatus func(EventHandlerCallRef nextHandler, EventRef anEvent, void *userData) {
  void (^block)() = (__bridge void (^)())(userData);
  block();
  return noErr;
}

static NSMutableArray *handlers;

+ (void)registerHotKey:(UInt32)keyCode modifiers:(UInt32)modifiers block:(nonnull void (^)())block;
{
  EventHotKeyRef hotKey;
  EventHotKeyID hotKeyID = {.signature = 'hush', .id = 1};
  EventTypeSpec eventType = {.eventClass = kEventClassKeyboard, .eventKind = kEventHotKeyPressed};

  void (^__nonnull block2)() = (__bridge void (^__nonnull)())Block_copy((__bridge void *)block);
  [handlers addObject:block2];
  InstallEventHandler(GetApplicationEventTarget(), func, 1, &eventType, (__bridge void *)block2, NULL);
  RegisterEventHotKey(keyCode, modifiers, hotKeyID, GetApplicationEventTarget(), kNilOptions, &hotKey);
}

@end

File: Hush/HotKeys.swift
import Cocoa
import Carbon

class HotKeys {
  static func registerHotKey(keyCode: UInt32, modifiers: UInt32, block: () -> ()) {
    let hotKeyID = EventHotKeyID(signature: 1, id: 1)
    var eventType = EventTypeSpec(eventClass: OSType(kEventClassKeyboard), eventKind: UInt32(kEventHotKeyPressed))

    let ptr = UnsafeMutablePointer<() -> ()>.alloc(1)
    ptr.initialize(block)

    InstallEventHandler(GetApplicationEventTarget(), {(_: EventHandlerCallRef, _: EventRef, ptr: UnsafeMutablePointer<Void>) -> OSStatus in
      UnsafeMutablePointer<() -> ()>(ptr).memory()
      return noErr
    }, 1, &eventType, ptr, nil)
    RegisterEventHotKey(keyCode, modifiers, hotKeyID, GetApplicationEventTarget(), OptionBits(0), nil)
  }
}
